### PR TITLE
enh: frontend to allow declaration of external function type after its initial declaration

### LINKED ISSUE #4495: llvm abort with external function
I have the following MRE derived from LAPACK's `src/sgedmd.f90`. The command to reproduce the error is `lfortran --implicit-interface --show-llvm -c err.f90` where err.f90 is the program containing the following code:
```
      PROGRAM SGEDMD

      USE                   iso_fortran_env
      IMPLICIT NONE

      REAL :: Y(1,1)

      EXTERNAL      ISAMAX
      INTEGER       ISAMAX
        
      IF ( Y(ISAMAX(2, Y(1,1),1),1 ) /= 0 ) THEN
        PRINT *, "X"
      END IF

      RETURN

      END 
```
gfortran compiles the program with `gfortran -c err.f90` whereas lfortran returns an error. There is no error when the conditional statment in the if-else is removed. The error only occurs during the ASR -> LLVM step.

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #4495.

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -2762,6 +2762,18 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                    /*
+                        return type of external function is specified
+                        external :: x
+                        integer :: x -> we are handling this case
+                    */
+                    ASR::symbol_t *sym_ = current_scope->get_symbol(sym);
+                    LCOMPILERS_ASSERT( sym_ != nullptr );
+                    // set function return type as `type`
+                    ASR::Function_t *f = ASR::down_cast<ASR::Function_t>(sym_);
+                    ASR::down_cast<ASR::Variable_t>(ASR::down_cast<ASR::Var_t>(f->m_return_var)->m_v)->m_type = type;
```
Comment (gxyd) [inline src/lfortran/semantics/ast_common_visitor.h:2775]: Can we maybe use `EXPR2VAR` for this? This looks a little complicated initially to understand.

**Comment by gxyd** [comment]
Looks good to me.

File: integration_tests/implicit_interface_17b.f90 [hunk 2]
```diff
@@ -0,0 +1,5 @@
+integer function isamax(x, y, z)
```
Comment (gxyd) [inline integration_tests/implicit_interface_17b.f90:1]: Curious, do we've the functionality in our test-suite to replicate `-c` (i.e.  to "Compile and assemble, do not link")?

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -2762,6 +2762,18 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                     is_allocatable, dims, type_declaration, s_abi,
                     (s_intent != ASRUtils::intent_local) || is_argument, is_dimension_star);
                 if ( is_attr_external ) create_external_function(sym, x.m_syms[i].loc, type);
+                if ( current_scope->get_symbol( sym ) != nullptr && ( is_external && !is_attr_external ) ) {
```
Comment (gxyd) [inline src/lfortran/semantics/ast_common_visitor.h:2765]: At this point, I lost faith in scrolling through `visit_DeclarationUtil`, it's such a big chunk now. 🗡️

File: integration_tests/implicit_interface_17b.f90 (hunk 2 above)
Comment (Pranavchiku) [reply integration_tests/implicit_interface_17b.f90:1]: yes, it is `-c`.

**Review by certik — APPROVED** [review]

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:2775]:
```suggestion
                    ASRUtils::EXPR2VAR(f->m_return_var)->m_type = type;
```

**Pull request merged**